lib/supavisor/db_handler.ex

```diff
@@ -255,4 +263,83 @@ defmodule Supavisor.DbHandler do
 
     :keep_state_and_data
   end
+
+  # Adapted from Postgrex.Protocol
+  defp handshake(sock, %{upstream_ssl: true} = auth), do: ssl(sock, auth)
+  defp handshake(sock, auth), do: send_startup(sock, auth)
+
+  defp ssl(sock, auth) do
+    case sock_send(sock, Server.ssl_request()) do
+      :ok -> ssl_recv(sock, auth)
+      {:disconnect, _, _} = dis -> dis
+    end
+  end
+
+  @spec ssl_recv(term(), map) ::
+          {:ok, {:ssl, :sslsock}}
+          | {:error, :ssl_not_available | :ssl_connect_error}
+  defp ssl_recv({:gen_tcp, sock} = s, auth) do
+    case :gen_tcp.recv(sock, 1, :infinity) do
+      {:ok, <<?S>>} ->
+        ssl_connect(s, auth)
+
+      {:ok, <<?N>>} ->
+        Logger.error("SSL requested but server says it's not available")
+        {:error, :ssl_not_available}
+
+      {:error, reason} ->
+        Logger.error("Error when receiving SSL response: #{inspect(reason)}")
+
+        {:error, :ssl_connect_error}
+    end
+  end
+
+  defp ssl_connect({:gen_tcp, sock}, auth, timeout \\ 5000) do
+    opts =
+      case auth.upstream_verify do
+        :peer ->
+          [
+            verify: :verify_peer,
+            cacerts: [auth.upstream_tls_ca],
```

Comment:
unless you have a strong opinion about using `ca/tls_ca` names, I'd consider using `cacerts/cacertfile` names because they'd be immediately obvious to most Elixir/Erlang people.

```suggestion
            cacerts: [auth.upstream_cacerts],
```

Reply:
I don't have a strong opinion about the naming. I just wanted to use something similar to [the PostgreSQL context](https://www.postgresql.org/docs/current/ssl-tcp.html)

<img width="1549" alt="Screenshot 2023-07-07 at 11 44 04" src="https://github.com/supabase/supavisor/assets/1172600/f76d00d5-9137-4165-9999-9011d3b10c09">

Reply:
in the next PRs, I'm gonna add `downstream_tls_cert`, `downstream_tls_key`, etc. WDYT?

Reply:
I see, sounds good to me!